cliotestutils/application.go

```diff
@@ -0,0 +1,73 @@
+package cliotestutils
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/anchore/clio"
+)
+
+// NewApplication takes a testing.T, a clio setup config, and a slice of assertions, and returns
+// a clio application that will, instead of setting up commands with their normal RunE, set up commands
+// such that the assertions are called with the testing.T after config state is set up by reading flags,
+// env vars, and config files. Useful for testing that expected configuration options are wired up.
+// Note that initializers will be cleared from the clio setup config, since the initialization may happen
+// more than once and affect global state. For necessary global state, a workaround is to set it in a TestingMain.
+func NewApplication(t *testing.T, cfg *clio.SetupConfig, assertions ...AssertionFunc) clio.Application {
+	cfg.Initializers = nil
```

Comment:
Note that we're not running the initializers here, because we can't guarantee they're safe to re-run. It's the caller's responsibility to set up global state outside the initializers for the test command. I would love for someone to have a better idea here.

Reply:
Could we wrap all the initializers in a `sync.Once` call or similar? I suppose it depends if we're calling `cliotestutil.NewApplication` multiple times to get multiple applications. Secondarily, these could have potentially different initializers. Or, we could also maintain some static state to indicate that the initializers ran, e.g. global:
```go
var stateInitialized = false
```
and around this line:
```go
originalInitializers := cfg.Initializers
cfg.Initializers = []clio.Initializer{ func(*clio.State) error {
  if stateInitialized {
    return nil
  }
  stateInitialized = true // might want some synchronization here, atomic.Bool wouldn't suffice
  // actually run originalInitializers
  return nil
}
```

Reply:
Thanks for the suggestion, but even if we use a `sync.Once` here it won't work, because NewApplication could be passed different configs at different times by different tests, and it would need a way to initialize each exactly once. So if we put a `sync.Once` on in the static space in this package, only the first application would have its initializers run, and if we put the `sync.Once` on the test application, global state can still be initialized twice by test callers making multiple applications. It might be possible to try to have some data structure that remembers which initializers have been run, but function pointers aren't comparable, so I'm not sure how that would work.